InvalidOperationException thrown while removing the subscribers in delete persistent subscription process. Modify the update persistent subscription strategy

dhingrak opened the pull request:
Fixed:  #3512  duplicate PS in the EventStoreDB UI

shaan1337 commented:
hey @dhingrak,

just wanting to make sure i understand the problem and solution correctly:

- my (rough) understanding after reading #3512 , is that the root cause is that saving the configuration to the config stream is done asynchronously and this opens up the possibility for a persistent subscription to be deleted in the _subscription* dictionaries but not yet in the config stream - so creating a new one could result in a duplicate. however, what's not clear to me, is that even if the above scenario occurred, _config should not be able to contain two subscriptions with the same name because the messages handled in the persistent subscription service are handled one at a time (see comment below about concurrency in solution section).

- to reproduce the issue, how did you simulate step 2 in #3512 ?
 
- we also have the issue where deleting/cleaning up the checkpoint and parked messages streams are done asynchronously and it may never actually happen if there is a timeout (but that's another problem)

solution:
- based on the code changes, it seems that you're relying on a lock to prevent a concurrent update (if i'm understanding it correctly). however a lock shouldn't be required since all the messages handled by the persistent subscriptions service are serialized in a queue and executed one at a time.

dhingrak commented:
@shaan1337 
Thanks a lot Shaan for your comments. Actually, I forgot to update the problem statement for https://github.com/EventStore/EventStore/issues/3512. 
After more investigation and discussion with Hayley we figured out it could be the create subscription part that are not handling the simultaneously request from clients.

I have tested the below scenario in my local environment. To reproduce this issue I have added a delay of 500 ms in createSubscriptionGroup section code and am able to create duplicate subscriptions. By adding a delay the check for existing persistent subscriptions passes when there are simultaneous requests and we have a duplicate subscription in UI. 

I have added a lock on the _subscriptionsById so the two clients are not able to update the state of _subscriptiionsById simultaneously to avoid the duplicate PS creation. 

**Update persistent subscription changes:** 

I have modify the code as suggested by Hayley so the update PS process completes in a single step rather than deleting the PS and then create a new PS

shaan1337 commented:
@dhingrak thanks for all the clarifications :+1: i'll give it a try

shaan1337 commented:
hey @dhingrak, i've checked the repro code you've posted in https://github.com/EventStore/EventStore/issues/3512

by adding `Task.Delay(...).ContinueWith...`, it is possible that the continuation will be run on another thread, which would cause the PersistentSubscriptionService to longer be thread-safe and could subsequently allow creation of duplicate persistent subscriptions.

would you be able to try with `Thread.Sleep(1000);` instead?

dhingrak commented:
hey @shaan1337 

I have tested the scenario with Thread.Sleep(1000) and I am not able to create duplicate PS. Indeed, the `Task.Delay(...).ContinueWith..` running on another thread in my local testing and that's why there are duplicate PS

dhingrak commented:
The above PR fixes the issue for InvalidOperationException thrown during the subscriber.shutdown process. Also, as suggested by Hayley we have modify the logic for updating the persistent subscriptions. Currently, the update process consists of delete and then recreate the persistent subscription. This PR will update the persistent subscription without deleting and recreating it.

shaan1337 commented on src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:
```diff
@@ -399,18 +402,19 @@ string user
 
 			if (!_subscriptionsById.ContainsKey(key)) {
-				onNotExist($"Group '{groupName}' does not exist.");
+				onNotExist($"Group '{groupName}' does not exists.");
```
```suggestion
				onNotExist($"Group '{groupName}' does not exist.");
```

dhingrak replied:
Done

shaan1337 replied:
thanks!

shaan1337 commented on src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:
```diff
@@ -626,22 +675,45 @@ string user
+
 			if (!_subscriptionsById.TryGetValue(key, out subscription)) {
-				onNotExist($"Group '{groupName}' does not exist.");
+				onNotExist($"Group '{groupName}' does not exists.");
```
```suggestion
				onNotExist($"Group '{groupName}' does not exist.");
```

dhingrak replied:
Done

shaan1337 commented on src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:
```diff
@@ -399,18 +402,19 @@ string user
 				return;
 			}
-
+			
```
```suggestion
```

shaan1337 commented on src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:
```diff
@@ -612,8 +647,22 @@ string user
 					new PersistentSubscriptionCheckpointWriter(key, _ioDispatcher),
 					new PersistentSubscriptionMessageParker(key, _ioDispatcher)));
+			
+			if (_subscriptionsById.ContainsKey(key)) {
```
suggestion: move this check before creating the PersistentSubscription object, otherwise it's not used.
it should also be ok to remove the Log.Information line since we don't currently log these errors:

```
			if (_subscriptionsById.ContainsKey(key))
				return false;
```

dhingrak replied:
Made the changes

shaan1337 replied:
thanks!

shaan1337 commented on src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:
```diff
@@ -612,8 +647,22 @@ string user
+			}
+
 			_subscriptionsById[key] = subscription;
+
```
suggestion:
move `List<PersistentSubscription> subscribers;` down here

dhingrak replied:
Done

shaan1337 commented on src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:
```diff
@@ -420,31 +424,55 @@ string user
+						break;
+					}
+				}
+			} else {
```
it seems that `subscription` is not being added to the `subscribers` list in the `else` part

actually, should we throw an exception in the `else` part? since the subscription should exist in `_subscriptionTopics` at this point.

shaan1337 replied:
edit: this may no longer be relevant if we go for the change suggested here: https://github.com/EventStore/EventStore/pull/3533/files#r987722268

shaan1337 commented on src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:
```diff
@@ -719,29 +791,6 @@ string user
 			);
 		}
 
-		private void RemoveSubscriptionConfig(string username, string eventSource, string groupName) {
```
suggestion:
to avoid code repetition in the update & remove methods, could we keep these two methods and add a parameter named: `replaceBy`  ?

```
RemoveSubscriptionConfig(string username, string eventSource, string groupName, PersistentSubscriptionEntry replaceBy);
RemoveSubscription(string eventSource, string groupName, PersistentSubscription replaceBy)
```

when `replaceBy` is null, we call RemoveAt(index) and when it's not null, we can replace the value at the index with that subscription or config.

shaan1337 replied:
the replacement null check would also need to be done for `_subscriptionsById.Remove(key);`

shaan1337 replied:
for more clarity, maybe we could then rename `RemoveSubscriptionConfig` to `UpdateSubscriptionConfig` and `RemoveSubscription` to `UpdateSubscription`

dhingrak replied:
Thanks a lot Shaan for the suggestions. Changes updated

shaan1337 replied:
thanks very much Kunal. I've made a few formatting suggestions and some additional refactoring suggestions for the creation part.

shaan1337 commented on src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:
```diff
@@ -460,10 +463,15 @@ string user
-			});
+			};
+			
+			UpdateSubscription(stream,groupName,subscription);
```
```suggestion
			UpdateSubscription(stream, groupName, subscription);
```

shaan1337 commented on src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:
```diff
@@ -718,24 +737,47 @@ string user
 			var index = _config.Entries.FindLastIndex(x => x.Stream == eventSource && x.Group == groupName);
-			_config.Entries.RemoveAt(index);
+			
+			// Updating the persistent subscription if replaceBy is not null, otherwise delete the persistent subscription
+			if (replaceBy != null) {
+				_config.Entries[index] = replaceBy;
+			} else {
+				_config.Entries.RemoveAt(index);
+			}
```
```suggestion
			// Update the persistent subscription if replaceBy is not null, otherwise delete the persistent subscription
			if (replaceBy != null)
				_config.Entries[index] = replaceBy;
			else
				_config.Entries.RemoveAt(index);
```

shaan1337 commented on src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:
```diff
@@ -718,24 +737,47 @@ string user
 			var key = BuildSubscriptionGroupKey(eventSource, groupName);
-			_subscriptionsById.Remove(key);
+			
+			// Updating the persistent subscription if replaceBy is not null, otherwise delete the persistent subscription
+			if(replaceBy != null)
+			{
+				_subscriptionsById[key] = replaceBy;
+			} else {
+				_subscriptionsById.Remove(key);
+			}
```
```suggestion
			// Update the persistent subscription if replaceBy is not null, otherwise delete the entry
			if (replaceBy != null)
				_subscriptionsById[key] = replaceBy;
			else
				_subscriptionsById.Remove(key);
```

shaan1337 commented on src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:
```diff
@@ -718,24 +737,47 @@ string user
+							Log.Error(ex, "Failed to shut down subscriber with id: {subscriptionId}", sub.SubscriptionId);
+						}
+
+						if (replaceBy != null) {
+							subscribers[i] = replaceBy;
+						} else {
+							subscribers.RemoveAt(i);
+						}
```
```suggestion
						if (replaceBy != null)
							subscribers[i] = replaceBy;
						else
							subscribers.RemoveAt(i);
```

shaan1337 commented on src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:
```diff
@@ -718,24 +737,47 @@ string user
+						try {
+							sub.Shutdown();
+						} catch (Exception ex) {
+							Log.Error(ex, "Failed to shut down subscriber with id: {subscriptionId}", sub.SubscriptionId);
```
```suggestion
							Log.Error(ex, "Failed to shut down subscription with id: {subscriptionId}", sub.SubscriptionId);
```

shaan1337 commented on src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:
```diff
@@ -399,18 +400,19 @@ string user
 				Action<string> onFail,
 				Action<string> onNotExist,
-				Action<string> onAccessDenied,
+				Action<string> onAccessDenied, 
```
```suggestion
				Action<string> onAccessDenied,
```

shaan1337 commented on src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:
```diff
@@ -583,13 +590,13 @@ string user
+			var stream = eventSource.ToString();
+			var key = BuildSubscriptionGroupKey(stream, groupName);
 
+			if (_subscriptionsById.ContainsKey(key)) {
+				return false;
+			}
```
```suggestion
			if (_subscriptionsById.ContainsKey(key))
				return false;
```

shaan1337 commented on src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:
```diff
@@ -253,11 +249,16 @@ string user
+			}
+
 			Log.Debug("New persistent subscription {subscriptionKey}", key);
 			_config.Updated = DateTime.Now;
```
similar to: https://github.com/EventStore/EventStore/pull/3533/files#r988645440
we can refactor this part and call `UpdateSubscriptionConfig(user, stream, groupName, createEntry);`

dhingrak commented:
Thanks a lot Shaan for helping me in refactoring the code.

shaan1337 commented on src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:
```diff
@@ -235,8 +231,8 @@ string user
-
-			CreateSubscriptionGroup(eventSource,
+			
+			var result = CreateSubscriptionGroup(eventSource,
```
suggestion:
1. we can rename `CreateSubscriptionGroup` to `TryCreateSubscriptionGroup` 
2. [here](https://github.com/EventStore/EventStore/blob/0f4a2db987faeec93d642c2dcc6cf7d63b54837f/src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs#L1028), we can log a warning when TryCreateSubscriptionGroup() returns false:
```
var key = BuildSubscriptionGroupKey(eventSource.ToString(), groupName);
Log.Warning("A duplicate persistent subscription: {subscriptionKey} was found in the configuration. Ignoring it.", key);
```
